pkg/controller/cloneset/sync/cloneset_update.go

```diff
@@ -187,16 +196,19 @@ func (c *realControl) updatePod(cs *appsv1alpha1.CloneSet, coreControl clonesetc
 
 		if c.inplaceControl.CanUpdateInPlace(oldRevision, updateRevision, coreControl.GetUpdateOptions()) {
 			switch state := lifecycle.GetPodLifecycleState(pod); state {
-			case "", appspub.LifecycleStateNormal:
-				var err error
-				var updated bool
-				if cs.Spec.Lifecycle != nil && lifecycle.IsPodHooked(cs.Spec.Lifecycle.InPlaceUpdate, pod) {
-					if updated, err = c.lifecycleControl.UpdatePodLifecycle(pod, appspub.LifecycleStatePreparingUpdate); err == nil && updated {
-						clonesetutils.ResourceVersionExpectations.Expect(pod)
-						klog.V(3).Infof("CloneSet %s update pod %s lifecycle to PreparingUpdate",
-							clonesetutils.GetControllerKey(cs), pod.Name)
+			case "", appspub.LifecycleStateNormal, appspub.LifecycleStateUpdated:
```

Comment:
How about keep the `appspub.LifecycleStateNormal` logic, add another `appspub.LifecycleStateUpdated`:

```golang
    case "", appspub.LifecycleStateNormal:
        // ...
    case appspub.LifecycleStateUpdated:
        var inPlaceUpdateHandler *appspub.LifecycleHook
        if cs.Spec.Lifecycle != nil {
            inPlaceUpdateHandler = cs.Spec.Lifecycle. InPlaceUpdate
        }
	if updated, err = c.lifecycleControl.UpdatePodLifecycleWithHandler(pod, appspub.LifecycleStatePreparingUpdate, inPlaceUpdateHandler); err == nil && updated {
		// ...
	}
	return 0, err
```

The `UpdatePodLifecycleWithHandler` function should patch pod with not only the lifecycle state, but also the finalizers and labels in the given handler.

This will also solve the problem in this PR that the pod object is changed in `SetPodLifecycleHook` but nowhere to update it to apiserver.

Reply:
done